packages/paste-website/src/components/assistant/Assistant.tsx

```diff
@@ -0,0 +1,143 @@
+
+  if (threadsStore == null) return null;
+
+  const handleMessageCreation = (message: string, threadId: string): void => {
+    // add the new user message to the store to optimistically render it whilst we wait for openAI to do its thing
+    addMessage(getMockMessage({ message }));
+
+    // Create a new "assistant run" on the thread so that openAI processes the new message and updates the thread with a response
+    createAssistantRun.mutate(
+      { threadId, message },
+      {
+        onSuccess: async (run) => {
+          // @ts-expect-error I don't know how to type this right now so it knows it's a response
+          const newRun = await run.json();
+          setActiveRun(newRun.run);
+        },
+      },
+    );
+
+    if (messages.length === 0) {
+      /**
+       * summarise the first question as the title of the
+       * thread, if this is the first message in that thread
+       * for it to appear in the thread list as an identifier
+       * of the thread
+       */
+      simpleCompletionMutation.mutate(
+        {
+          prompt:
+            "If this is the start of a threaded conversation, summarise the subject of the converation in as few words as possible: ",
+          context: message,
+        },
+        {
+          onSuccess: async (completion) => {
+            // @ts-expect-error I don't know how to type this right now so it knows it's a response
+            const newCompletion = await completion.json();
+            // update the thread title in the store
+            threadsStore?.setThreadTitle(threadId, newCompletion.choices[0].message.content);
+            // update the thread title in openAI
+            updateThreadMutation.mutate({ id: threadId, threadTitle: newCompletion.choices[0].message.content });
+          },
+        },
+      );
+    }
+  };
+
+  /**
+   * From one of the canned new thread messages, create a new thread, select it, and then
+   * create a new message run on the newly created thread.
+   *
+   * @param {string} message
+   */
+  const handleCannedThreadCreation = (message: string): void => {
+    createThreadMutation.mutate(
+      {},
+      {
+        onSuccess: async (data) => {
+          // @ts-expect-error I don't know how to type this right now so it knows it's a response
+          const newThread = await data.json();
+          if (threadsStore == null) return;
+          threadsStore.createAndSelectThread(newThread);
+          handleMessageCreation(message, newThread.id);
+        },
+      },
+    );
+  };
+
+  return (
+    <AsssistantLayout.Window>
+      <AsssistantLayout.Threads>
+        <AsssistantLayout.ThreadsHeader>
+          <AssistantHeader />
+        </AsssistantLayout.ThreadsHeader>
+        <AssistantThreads />
+      </AsssistantLayout.Threads>
+      <AsssistantLayout.Canvas>
+        {threadsStore.selectedThreadID == null && (
+          <AssistantEmptyState onCannedThreadCreation={handleCannedThreadCreation} />
```

Comment:
So I pulled it up to this level and passed it down because both `handleCannedThreadCreation` and `handleMessageCreation` mutate the same state. 

When they both handled their own mutations in `AssistantEmptyState` and `AssistantComposer` respectively, it was causing annoying and weird re-renders

Reply:
Understood.